proctor-consumer/src/main/java/com/indeed/proctor/consumer/ProctorConsumerUtils.java

```diff
@@ -106,16 +108,26 @@ public static String getForceGroupsStringFromRequest(@Nonnull final HttpServletR
             return "";
         }
 
-        for (int i = 0; i < cookies.length; i++) {
-            if (FORCE_GROUPS_COOKIE_NAME.equals(cookies[i].getName())) {
-                final String cookieValue = cookies[i].getValue();
-                return Strings.nullToEmpty(cookieValue);
-            }
-        }
-
-        return "";
+        return Arrays.stream(cookies).sequential()
+                .filter(Objects::nonNull)
+                .filter(x -> FORCE_GROUPS_COOKIE_NAME.equals(x.getName()))
+                .sorted(COOKIE_COMPARATOR)
+                .map(Cookie::getValue)
+                .filter(Objects::nonNull)
+                .collect(Collectors.joining(","));
     }
 
+    /**
+     * It's not enough to just combine all the force groups from all the available cookies. If different force group
+     * cookies exist with different allocations for the same test, we could have different results. We need to process
+     * the cookies in a stable order so that we always produce the same result.
+     */
+    @VisibleForTesting
+    private static final Comparator<Cookie> COOKIE_COMPARATOR = Comparator
+            .comparing(Cookie::getPath)
+            .thenComparing(Cookie::isHttpOnly)
+            .thenComparing(Cookie::getValue);
```

Comment:
> But we're not doing that. We are taking the first cookie that matches the given name. According to this spec, an cookie on ".indeed.net" will be listed first in the Cookies header before a cookie on "www.indeed.net", regardless of the age.

Right. I think the current implementation assumes (or works well only if) there is at most one prforceGroups cookie in a request.

By the way, it's worth double-checking but I think clients don't send cookie's `path` and `isHttpOnly` values to server because they're used only in `Set-Cookie` context. Please check https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cookie#syntax for the `Cookie` syntax.

So actually we have no ways to sort cookies. So I'd suggest keeping the order as it is.

Reply:
> Right. I think the current implementation assumes (or works well only if) there is at most one prforceGroups cookie in a request.

Sure. But that's absolutely 100% not the case. And the fact that we have more than one prforceGroups cookie in a request is thought to be the cause of a persistent problem that's interfering with testing a major feature.

You're right that the browser doesn't send `path` or `isHttpOnly`. But we can still do something by merging the groups. The order will (probably) be consistent. It may cause problems, but I expect it to cause _fewer_ problems.

Considering this has been a present and very frustrating bug that is actually impeding in our ability to get stuff done, and because the RFC explicitly says you shouldn't rely on the ordering cookies appear in the header, I offer a few proposals:

1. We take the list of cookies that match the name, reverse the order, and join the cookie values together. So you'll get all the distinct tests forced from all your cookies, but the ones that are first in the list will win any conflicts between groups.
2. Take the list of cookies and sort by some other value (such as the number of groups forced or the a simple string comparison on the prforceGroups cookie) and choose the one that sorts first.
3. Take the list of cookies and sort by some other value (such as the number of groups forced or the a simple string comparison on the prforceGroups cookie) and join the cookie values together. Like #1 but we don't rely on the ordering the browser uses in the Cookie header.
4. Update the mechanisms in writing the force groups cookie to try and avoid the possibility of multiple groups writing cookies on paths that would overlap. For example, we may be able to find a mechanism to prevent somebody from setting the cookie on ".example.com" if their context path is "www.example.com".
5. Encode "domain" and/or "path" field in the cookie payload to help with the sorting of which cookie to use and to better diagnose which applications are setting cookies on paths for which they shouldn't.
6. Encode "last updated field" in the cookie payload and choose the one with the latest last updated date. Break ties either by order in the header or by any other order. This is an order of magnitude more effort, and it would take more time to implement than is probably feasible for this effort. So I'm willing to do this a few weeks from now but in the meantime I'd like to deploy some other fix to unblock first.

Reply:
Thanks for giving the proposals. To clarify, I'm not against merging prforceGroups cookies. I just suggest not sorting them and keeping the given order. By merging prforceGroups cookie values without sorting, I think we get the same result as your suggestion 1.

Reply:
Oddly, I am seeing some evidence that the Chrome isn't actually upholding this order.

1.  Use an extension to set a prforceGroups cookie on ".indeed.net" with value "badtst_1"
2. Visit "www.qa.indeed.net/?prforceGroups=goodtst_1"
3. Check your cookies. You'll see a prforceGroups cookie on ".indeed.net" and "www.qa.indeed.net". Both have a value of "badtst_1".

In this case the homepage request will correctly set the cookie to goodtst_1, but an AJAX call in that same session to "www.qa.indeed.net/hp/rpc/frontendlogging" puts them in the order of "badtst_1" _then_ "goodtst_1", so goodtst_1 is getting overwritten.

Not a deal breaker for this issue, but weirdness at which to continue poking.

Reply:
Ah, I think I understand. The `path` fields for these cookies are the same; both have a path of "/". Which is to say, I was assuming the domain was part of the path, but it's not. They're separate. So it's sorting purely on date. So doing a reverse order search would be wrong.

Reply:
Thanks for updating!